### PR TITLE
Cache `Site.get_site_root_paths` in the request scope (v2)

### LINKED ISSUE #2916: Wagtail can make MANY requests to wagtail_site_root_paths
One site I'm building is making over 80 requests for the cache key `wagtail_site_root_paths`, which I hadn't noticed until I put a proper caching engine in place and suddenly performance actually dropped due to the round trips.

Perhaps replace all calls to `cache.get('wagtail_site_root_paths')` with a memoized result, so that we're making at most 1 request for this key?

### PR DESCRIPTION
**Pull request by tobiasmcnulty**
Fixes #2916
RFC: [#14](https://github.com/tobiasmcnulty/rfcs/blob/master/draft/0014-page-url-api.rst)

This is a rebased and updated version of #3188 by @gasman that moves the business logic for caching `site_root_paths` inside of the `Page` model. This makes it easier to use the API from outside Wagtail itself (the most common use case for this is most likely any navigation menu-related template tags, e.g., https://github.com/wagtail/wagtaildemo/pull/110/files).

- [x] Add/update usage documentation/recommendations for using and overriding URL-related methods on the `Page` model, if this is the route we want to go
- [x] Determine whether to deprecate `Page.url` and `Page.full_url` or defer that decision till a later date. While convenient, using them circumvents the possibility of caching `site_root_paths` and is probably a habit worth disincentivizing. (DECISION: don't deprecate)
- [x] Determine if passing `request` to these methods is appropriate (seems to be okay, other methods on `Page` already accept a `request`)

See also: #3350

### CODE REVIEW

File: docs/topics/pages.rst [hunk 1]
```diff
@@ -192,6 +192,47 @@ By default, any page type can be created under any page type and it is not neces
...
+
+While you could pass only the ``request`` keyword argument, passing all arguments as-is ensures compatibility with any
+future changes to these method signatures.
+
+For more information, please see :meth:`wagtail.wagtailcore.models.Page.get_url_parts`.
+
+Obtaining URLs for page instances
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The ``Page.get_url(request)`` method can be called whenever a page URL is needed. It defaults to returning local URLs (not including the protocol or domain) if it can detect that the page is on current site (via ``request.site``); otherwise, a full URL including the protocal and domain is returned. Whenever possible, the optional ``request`` argument should be included to enable per-request caching of site-level URL information and facilitate the generation of local URLs.
```
Comment (gasman) [inline docs/topics/pages.rst]: protocal -> protocol

File: docs/topics/pages.rst [hunk 2]
```diff
@@ -192,6 +192,47 @@ By default, any page type can be created under any page type and it is not neces
...
+
+A common use case for ``get_url(request)`` is in any custom template tag your project may include for generating navigation menus. When writing a such a
+custom template tag, ensure it includes ``takes_context=True`` and use ``context.get('request')`` to safely pass the
+request or ``None`` if no request exists in the context.
+
+For more information, please see :meth:`wagtail.wagtailcore.models.Page.get_url`.
+
+In the event a full URL (including the protocol and domain is needed), ``Page.get_full_url(request)`` can be used instead. Whenever possible, the optional ``request`` argument should be included to enable per-request caching of site-level URL information. For more information, please see :meth:`wagtail.wagtailcore.models.Page.get_full_url`.
```
Comment (gasman) [inline docs/topics/pages.rst]: Closing bracket should be after 'domain': a full URL(including the protocol and domain) is needed

File: wagtail/wagtailcore/models.py [hunk 3]
```diff
@@ -32,11 +33,12 @@
 from treebeard.mp_tree import MP_Node
 
 from wagtail.utils.compat import user_is_authenticated
+from wagtail.utils.deprecation import RemovedInWagtail110Warning
```
Comment (gasman) [inline wagtail/wagtailcore/models.py]: This should now be `RemovedInWagtail112Warning`

File: wagtail/wagtailcore/models.py [hunk 4]
```diff
@@ -596,6 +598,18 @@ def check(cls, **kwargs):
                 )
             )
 
+        if not accepts_kwarg(cls.relative_url, 'request'):
+            warnings.warn(
+                "%s.relative_url should accept a 'request' keyword argument. "
+                "See http://docs.wagtail.io/en/v1.8/reference/pages/model_reference.html#wagtail.wagtailcore.models.Page.relative_url" % cls,
```
Comment (gasman) [inline wagtail/wagtailcore/models.py]: This link (and the one below) will now need to point to the (as-yet-unpublished) `v1.10` docs.

File: wagtail/wagtailcore/models.py [hunk 5]
```diff
@@ -784,7 +798,33 @@ def is_navigable(self):
...
+            # request is None OR the cache hasn't been set yet
+            site_root_paths = Site.get_site_root_paths()
+            if request is not None:
+                request._wagtail_cached_site_root_paths = site_root_paths
+        return site_root_paths
+
+    def _safe_get_url_parts(self, request=None):
+        """
+        Backwards-compatibility method to safely call get_url_parts without
+        the new ``request`` kwarg (added in Wagtail 1.11), if needed.
```
Comment (gasman) [inline wagtail/wagtailcore/models.py]: Should be Wagtail 1.10

File: wagtail/wagtailcore/templatetags/wagtailcore_tags.py [hunk 6]
```diff
@@ -18,7 +19,22 @@ def pageurl(context, page):
...
     current page, or absolute (http://example.com/foo/bar/) if not.
     """
-    return page.relative_url(context['request'].site)
+    try:
+        current_site = context['request'].site
+    except (KeyError, AttributeError):
+        # request.site not available in the current context; fall back on page.url
+        return page.url
+
+    # RemovedInWagtail110Warning - this accepts_kwarg test can be removed when we drop support
```
Comment (gasman) [inline wagtail/wagtailcore/templatetags/wagtailcore_tags.py]: RemovedInWagtail112Warning

**Comment by tobiasmcnulty** [comment]
Hi @gasman, sorry for the delay on this. I rebased on master again and removed the deprecation TODO items. Does this look all right to you (assuming Travis CI passes)? And would you prefer to merge with the history intact, or would you like me to squash this into a single commit?

Thanks!

**Comment by tobiasmcnulty** [comment]
Hi @gasman, I've rebased this on master again and moved the release notes to 1.11. Please let me know if this looks good or if you see any tweaks I could make.

**Comment by gasman** [comment]
Squashed / merged in 1f8edc8984beac8040569c2db3c7ad0cad06f550 - thanks @tobiasmcnulty!

**Pull request closed without merging**